Port formatChatMessage to C++ for internationalization

poikilos opened the pull request:
- Goal of the PR
  Allow international strings (with 2-byte characters) in formatted chat messages.
- How does the PR work?
  1. Move `format_chat_message` functionality to the existing C++ `formatChatMessage` function.
  2. Make the `formatChatMessage` use `wstring` throughout, except for getting substrings that are only available as `string` (the rest of the `wstring` remains intact even in that case).
  3. Make an overload for getTimestamp (to allow, in this case, the format `"%H:%M:%S"` which was the format used by the Lua function).
- Does it resolve any reported issue?
  - reported by @MoNTE48  Cyrillic text is lost in chat due to calling `wide_to_narrow`, on Android 4, some other newer Android devices, and possibly a legacy OS.
- If not a bug fix, why is this PR needed? What use cases does it solve?
  - Fix bugs that started in:
    - https://github.com/minetest/minetest/pull/9133
    - https://github.com/minetest/minetest/pull/8529


## To do

This PR is Ready for Review.

- Suggested future options: If future functionality requires Lua to completely change formatChatMessage (as opposed to using only the minetest.conf setting `chat_message_format`), which seems unnecessary (and may be why format_chat_message was in Lua before, according to the vague code comment), Lua should call this function, not the other way around like before (to preserve international strings on all platforms).

## How to test
- Your computer (or Android device) must be set to Cyrillic text (Russian).
- Paste "русский текст" into chat
- Check the log
  - behavior before PR: Cyrillic characters are discarded in the GUI and in the log on some systems (verified on Android 4).
  - behavior after PR: Cyrillic characters remain intact (verified on Android 4)

sfan5 commented on src/gettime.h:
```diff
@@ -30,13 +30,18 @@ enum TimePrecision
 	// This is not really thread-safe but it won't break anything
 	// except its own output, so just go with it.
 	struct tm *tm = localtime(&t);
 	char cs[20]; // YYYY-MM-DD HH:MM:SS + '\0'
```
buffer size needs adjustment too

sfan5 commented on src/util/string.h:
```diff
@@ -473,6 +473,23 @@ inline void str_replace(std::string &str, const std::string &pattern,
+		const std::wstring &replacement)
+{
+	std::wstring::size_type start = str.find(pattern, 0);
+	while (start != str.npos) {
```
str.npos looks weird
consider `std::wstring::npos` instead

poikilos replied:
I am giving the change to MoNTE48, and changes to the instances that were already there as well.

poikilos replied:
This is copied and pasted from the existing function--the standard library probably puts npos in the object just for that reason (it can avoid copy&paste errors when people make future changes as well). The file uses `str.npos` throughout, with or without the PR.

sfan5 commented on src/script/cpp_api/s_server.cpp:
```diff
@@ -168,25 +171,26 @@ void ScriptApiServer::on_shutdown()
 
-std::string ScriptApiServer::formatChatMessage(const std::string &name,
-	const std::string &message)
+std::wstring ScriptApiServer::formatChatMessage(const std::wstring &name,
```
this could be moved out of `ScriptApiServer` now that it no longer calls Lua

poikilos replied:
Where to?

rubenwardy commented:
I asked for the function to be moved from C++ to Lua as it makes it much easier for mods to change the chat message formatting (for example: colorising nicks by teams). If it's not possible to do internationalization in Lua, then that indicates a deficiency in the Lua API

poikilos commented:
> I asked for the function to be moved from C++ to Lua as it makes it much easier for mods to change the chat message formatting (for example: colorising nicks by teams). If it's not possible to do internationalization in Lua, then that indicates a deficiency in the Lua API. If mods can't encode characters in chat messages correctly then there is something very wrong
> 
> Also, this is currently a breaking change as you're removing an API function

If there is something really wrong, and there is no other solution on those platforms, we can make this function accessible to Lua, re-add format_chat_message, but make it call this function. How about that?

SmallJoker commented:
```Lua
function core.format_chat_message()
    minetest.log("warning", "minetest.format_chat_message is obsolete!")
end
```

rubenwardy commented:
> If there is something very wrong, and there is no other solution on those platforms, we can make this function accessible to Lua, re-add format_chat_message, but make it call this function. How about that?

If there's something very wrong, then we might as well abandon all hope. If Lua breaks internationalization, then we shouldn't have used it.

I'm not the most experienced with string encoding, but I know that wstring isn't necessary everywhere. You can get the same behaviour with UTF-8, you just need to make sure to do conversions correctly

poikilos commented:
@rubenwardy MoNTE48 helped me talk with gmax79 in Russian, who also knows English we discovered (unless he is using a translator). gmax79 says that his custom lua_pushwsting function is just for convenience--it merely converts the string to UTF-8 then passes it off to lua_pushstring. He doesn't have an Android build so there is no way to check for the problem. MoNTE48 and I are working on some more theories as to why the problem occurs. I'll see if there is an easier solution than this that I can implement for him.

nerzhul commented on src/gettime.h:
```diff
@@ -30,13 +30,18 @@ enum TimePrecision
 };
 
-inline std::string getTimestamp()
+inline std::string getTimestamp(std::string fmt)
```
const char * please, we don't need c++ there it's a const raw C parameter

poikilos replied:
I'll make the change and send it to MoNTE48.

poikilos replied:
This is changed now.

SmallJoker commented on src/util/string.h:
```diff
@@ -473,6 +473,23 @@ inline void str_replace(std::string &str, const std::string &pattern,
++ * @param pattern The pattern to replace.
++ * @param replacement What to replace the pattern with.
++ */
+inline void str_replace(std::wstring &str, const std::wstring &pattern,
```
You could make this function return the amount of replaced parts (amount of loops, so to say).
This way it your string duplication and `compare()` call would become superfluous.

Just as an idea, because I think that's a cleaner solution.

MoNTE48 replied:
Just take a look at line 466

rubenwardy commented:
:-1: for this PR. Please fix the encoding conversion to and from Lua instead

poikilos commented:
I've learned that UTF8 supports encoding of a variable number of bytes, despite its name, my mistake. That means Minetest should be able to handle Asian text maybe with changes. I have no way of testing this, but I suppose I could put Ubuntu Kylin in a VM and how far I get. I have a friend who can read some Chinese––Does it work currently? Near the slow code that I fixed that was looping on every character and which I changed to instead use a `std::map`: 1-byte character filtering in Minetest (for "security" I suppose), would seem to break this, explain why not if not. Related: issues #373 (Irrlicht bug fixed but not in an Irrlicht release yet) #1632 #1682 #2574, and focus of the issue, according to 2574, is #2885 . See also https://github.com/minetest/minetest/pull/9117

sfan5 commented:
When you have Minetest's language configured to one that enables the fallback font (incl. Chinese), Chinese characters written in chat will display:

![screenshot_20201211_131729](https://user-images.githubusercontent.com/1042418/101902836-992d4e00-3bb3-11eb-9646-cdf9298fd211.png)

What remains broken is:
* display of Chinese text (and other certain languages) when the fallback font is *not* enabled
* copy-paste of Unicode text (all platforms?)
* utf-8 conversion issues on Android, Windows(?) and Mac(?)
